File: php/src/VedaLoadBalancer.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Load balancer across multiple VedaDB nodes.
 */
class VedaLoadBalancer
{
    public const STRATEGY_ROUND_ROBIN = 'round_robin';
    public const STRATEGY_RANDOM      = 'random';
    public const STRATEGY_LEAST_CONN  = 'least_connections';
    public const STRATEGY_WEIGHTED    = 'weighted';
    public const STRATEGY_HASH        = 'hash';

    /** @var list<array{client: VedaClient, weight: int, healthy: bool, connections: int}> */
    private array $nodes = [];

    private string $strategy;
    private int $currentIndex = 0;
    private int $healthCheckIntervalMs;
    private ?float $lastHealthCheck = null;

    public function __construct(
        string $strategy = self::STRATEGY_ROUND_ROBIN,
        int $healthCheckIntervalMs = 30000,
    ) {
        $this->strategy                = $strategy;
        $this->healthCheckIntervalMs   = $healthCheckIntervalMs;
    }

    /**
     * Add a node to the pool.
     */
    public function addNode(VedaClient $client, int $weight = 1): void
    {
        $this->nodes[] = [
            'client'      => $client,
            'weight'      => max(1, $weight),
            'healthy'     => true,
            'connections' => 0,
        ];
    }

    /**
     * Remove a node from the pool.
     */
    public function removeNode(VedaClient $client): void
    {
        $oid = spl_object_id($client);
        $this->nodes = array_values(array_filter(
            $this->nodes,
            fn(array $n) => spl_object_id($n['client']) !== $oid,
        ));
    }

    /**
     * Get the next node based on strategy.
     */
    public function nextNode(): VedaClient
    {
        $this->maybeHealthCheck();

        $healthyNodes = array_filter($this->nodes, fn(array $n) => $n['healthy']);
        if (empty($healthyNodes)) {
            throw new FailoverException('No healthy nodes available');
        }

        $node = match ($this->strategy) {
            self::STRATEGY_ROUND_ROBIN => $this->pickRoundRobin($healthyNodes),
            self::STRATEGY_RANDOM      => $this->pickRandom($healthyNodes),
            self::STRATEGY_LEAST_CONN  => $this->pickLeastConnections($healthyNodes),
            self::STRATEGY_WEIGHTED    => $this->pickWeighted($healthyNodes),
            self::STRATEGY_HASH        => $this->pickRoundRobin($healthyNodes),
            default                    => $this->pickRoundRobin($healthyNodes),
        };

        $node['connections']++;
        return $node['client'];
    }

    private function pickRoundRobin(array $nodes): array
    {
        $values = array_values($nodes);
        $picked = $values[$this->currentIndex % count($values)];
        $this->currentIndex++;
        return $picked;
    }

    private function pickRandom(array $nodes): array
    {
        $values = array_values($nodes);
        return $values[array_rand($values)];
    }

    private function pickLeastConnections(array $nodes): array
    {
        $min = PHP_INT_MAX;
        $picked = null;
        foreach ($nodes as $node) {
            if ($node['connections'] < $min) {
                $min = $node['connections'];
                $picked = $node;
            }
        }
        return $picked ?? array_values($nodes)[0];
    }

    private function pickWeighted(array $nodes): array
    {
        $totalWeight = array_sum(array_map(fn(array $n) => $n['weight'], $nodes));
        $random = mt_rand(1, max(1, $totalWeight));

        $cumulative = 0;
        foreach ($nodes as $node) {
            $cumulative += $node['weight'];
            if ($random <= $cumulative) {
                return $node;
            }
        }

        return array_values($nodes)[0];
    }

    /**
     * Release a connection on a node.
     */
    public function releaseNode(VedaClient $client): void
    {
        foreach ($this->nodes as &$node) {
            if (spl_object_id($node['client']) === spl_object_id($client)) {
                $node['connections'] = max(0, $node['connections'] - 1);
                return;
            }
        }
    }

    /**
     * Execute a callback on a selected node.
     *
     * @template T
     * @param callable(VedaClient): T $fn
     * @return T
     */
    public function execute(callable $fn): mixed
    {
        $node = $this->nextNode();
        try {
            return $fn($node);
        } finally {
            $this->releaseNode($node);
        }
    }

    /**
     * Run health checks if interval has elapsed.
     */
    public function maybeHealthCheck(): void
    {
        if ($this->lastHealthCheck === null) {
            $this->healthCheck();
            return;
        }

        $elapsed = (microtime(true) * 1000) - $this->lastHealthCheck;
        if ($elapsed >= $this->healthCheckIntervalMs) {
            $this->healthCheck();
        }
    }

    /**
     * Run health checks on all nodes.
     *
     * @return list<array{node_index: int, healthy: bool}>
     */
    public function healthCheck(): array
    {
        $results = [];
        foreach ($this->nodes as $i => &$node) {
            try {
                $healthy = $node['client']->ping();
            } catch (\Throwable) {
                $healthy = false;
            }
            $node['healthy'] = $healthy;
            $results[] = ['node_index' => $i, 'healthy' => $healthy];
        }

        $this->lastHealthCheck = microtime(true) * 1000;
        return $results;
    }

    /**
     * Get all healthy nodes.
     *
     * @return list<VedaClient>
     */
    public function getHealthyNodes(): array
    {
        return array_values(array_filter(
            array_map(fn(array $n) => $n['healthy'] ? $n['client'] : null, $this->nodes),
        ));
    }

    /**
     * Get node count.
     */
    public function getNodeCount(): int
    {
        return count($this->nodes);
    }

    /**
     * Get healthy node count.
     */
    public function getHealthyCount(): int
    {
        return count(array_filter($this->nodes, fn(array $n) => $n['healthy']));
    }

    /**
     * Get node metrics.
     */
    public function getMetrics(): array
    {
        $nodes = [];
        foreach ($this->nodes as $i => $node) {
            $nodes[] = [
                'index'       => $i,
                'weight'      => $node['weight'],
                'healthy'     => $node['healthy'],
                'connections' => $node['connections'],
            ];
        }

        return [
            'strategy'        => $this->strategy,
            'total_nodes'     => count($this->nodes),
            'healthy_nodes'   => $this->getHealthyCount(),
            'nodes'           => $nodes,
        ];
    }
}

File: dotnet/VedaLoadBalancer.cs
using System.Collections.Concurrent;
using System;

namespace VedaDB;

/// <summary>
/// Strategy for selecting a node from the pool.
/// </summary>
public enum LoadBalanceStrategy
{
    /// <summary>Round-robin selection.</summary>
    RoundRobin,
    /// <summary>Random selection.</summary>
    Random,
    /// <summary>Least connections.</summary>
    LeastConnections,
    /// <summary>Weighted round-robin.</summary>
    WeightedRoundRobin,
    /// <summary>IP hash for sticky sessions.</summary>
    IPHash
}

/// <summary>
/// A node in the load balancer pool.
/// </summary>
public class VedaNode
{
    /// <summary>
    /// Node host.
    /// </summary>
    public string Host { get; set; } = "";

    /// <summary>
    /// Node port.
    /// </summary>
    public int Port { get; set; }

    /// <summary>
    /// Node weight (for weighted strategies).
    /// </summary>
    public int Weight { get; set; } = 1;

    /// <summary>
    /// Whether the node is currently healthy.
    /// </summary>
    public bool IsHealthy { get; set; } = true;

    /// <summary>
    /// Current number of active connections.
    /// </summary>
    public int ActiveConnections;

    /// <summary>
    /// Total requests handled.
    /// </summary>
    public long TotalRequests;

    /// <summary>
    /// Node identifier.
    /// </summary>
    public string Id => $"{Host}:{Port}";
}

/// <summary>
/// Load balancer for distributing requests across multiple VedaDB nodes.
/// </summary>
public class VedaLoadBalancer
{
    private readonly LoadBalanceStrategy _strategy;
    private readonly List<VedaNode> _nodes = new();
    private readonly object _nodeLock = new();
    private int _roundRobinIndex;
    private readonly Random _random = new();

    /// <summary>
    /// Create a load balancer.
    /// </summary>
    public VedaLoadBalancer(LoadBalanceStrategy strategy = LoadBalanceStrategy.RoundRobin)
    {
        _strategy = strategy;
    }

    /// <summary>
    /// Add a node to the pool.
    /// </summary>
    public void AddNode(string host, int port, int weight = 1)
    {
        lock (_nodeLock)
        {
            _nodes.Add(new VedaNode { Host = host, Port = port, Weight = weight });
        }
    }

    /// <summary>
    /// Remove a node from the pool.
    /// </summary>
    public void RemoveNode(string host, int port)
    {
        lock (_nodeLock)
        {
            _nodes.RemoveAll(n => n.Host == host && n.Port == port);
        }
    }

    /// <summary>
    /// Update node health status.
    /// </summary>
    public void SetNodeHealth(string host, int port, bool isHealthy)
    {
        lock (_nodeLock)
        {
            var node = _nodes.FirstOrDefault(n => n.Host == host && n.Port == port);
            if (node != null) node.IsHealthy = isHealthy;
        }
    }

    /// <summary>
    /// Get all registered nodes.
    /// </summary>
    public IReadOnlyList<VedaNode> GetNodes()
    {
        lock (_nodeLock) { return _nodes.ToList(); }
    }

    /// <summary>
    /// Get only healthy nodes.
    /// </summary>
    public List<VedaNode> GetHealthyNodes()
    {
        lock (_nodeLock) { return _nodes.Where(n => n.IsHealthy).ToList(); }
    }

    /// <summary>
    /// Select the next node based on the configured strategy.
    /// </summary>
    public VedaNode? SelectNode(string? clientKey = null)
    {
        var healthyNodes = GetHealthyNodes();
        if (healthyNodes.Count == 0) return null;

        VedaNode? selected;

        switch (_strategy)
        {
            case LoadBalanceStrategy.Random:
                selected = healthyNodes[_random.Next(healthyNodes.Count)];
                break;

            case LoadBalanceStrategy.LeastConnections:
                selected = healthyNodes.OrderBy(n => n.ActiveConnections).First();
                break;

            case LoadBalanceStrategy.WeightedRoundRobin:
                selected = WeightedSelection(healthyNodes);
                break;

            case LoadBalanceStrategy.IPHash:
                selected = IPHashSelection(healthyNodes, clientKey);
                break;

            case LoadBalanceStrategy.RoundRobin:
            default:
                selected = RoundRobinSelection(healthyNodes);
                break;
        }

        if (selected != null)
        {
            selected.TotalRequests++;
            selected.ActiveConnections++;
            VedaMetrics.Increment("vedadb_loadbalancer_selects", 1,
                new() { { "node", selected.Id }, { "strategy", _strategy.ToString() } });
        }

        return selected;
    }

    /// <summary>
    /// Release a connection back to the pool, decrementing active count.
    /// </summary>
    public void ReleaseNode(VedaNode node)
    {
        node.ActiveConnections--;
    }

    /// <summary>
    /// Execute an operation on a selected node.
    /// </summary>
    public async Task<T> ExecuteAsync<T>(Func<VedaNode, Task<T>> operation, string? clientKey = null)
    {
        var node = SelectNode(clientKey);
        if (node == null)
            throw new VedaConnectionException("No healthy nodes available in load balancer");

        try
        {
            return await operation(node);
        }
        catch (Exception ex)
        {
            VedaMetrics.Increment("vedadb_loadbalancer_errors", 1,
                new() { { "node", node.Id }, { "error", ex.GetType().Name } });
            throw;
        }
        finally
        {
            ReleaseNode(node);
        }
    }

    /// <summary>
    /// Get pool statistics.
    /// </summary>
    public VedaLoadBalancerStats GetStats()
    {
        var nodes = GetNodes();
        return new VedaLoadBalancerStats
        {
            TotalNodes = nodes.Count,
            HealthyNodes = nodes.Count(n => n.IsHealthy),
            UnhealthyNodes = nodes.Count(n => !n.IsHealthy),
            TotalRequests = nodes.Sum(n => n.TotalRequests),
            ActiveConnections = nodes.Sum(n => n.ActiveConnections)
        };
    }

    private VedaNode? RoundRobinSelection(List<VedaNode> nodes)
    {
        if (nodes.Count == 0) return null;
        var index = Interlocked.Increment(ref _roundRobinIndex);
        return nodes[Math.Abs(index) % nodes.Count];
    }

    private VedaNode? WeightedSelection(List<VedaNode> nodes)
    {
        if (nodes.Count == 0) return null;
        var totalWeight = nodes.Sum(n => n.Weight);
        var pick = _random.Next(totalWeight);
        var current = 0;
        foreach (var node in nodes)
        {
            current += node.Weight;
            if (pick < current) return node;
        }
        return nodes.Last();
    }

    private VedaNode? IPHashSelection(List<VedaNode> nodes, string? clientKey)
    {
        if (nodes.Count == 0) return null;
        var key = clientKey ?? Guid.NewGuid().ToString();
        var hash = key.GetHashCode();
        return nodes[Math.Abs(hash) % nodes.Count];
    }
}

/// <summary>
/// Statistics for the load balancer.
/// </summary>
public class VedaLoadBalancerStats
{
    public int TotalNodes { get; set; }
    public int HealthyNodes { get; set; }
    public int UnhealthyNodes { get; set; }
    public long TotalRequests;
    public int ActiveConnections;
}
